File: app/src/main/java/com/qadam/lastpuff/data/repository/UserRepository.kt
package com.qadam.lastpuff.data.repository

import com.qadam.lastpuff.data.datastore.PreferencesManager
import com.qadam.lastpuff.data.local.dao.CravingEventDao
import com.qadam.lastpuff.data.local.dao.MoneyGoalDao
import com.qadam.lastpuff.data.local.dao.RelapseEventDao
import com.qadam.lastpuff.data.local.dao.SosContactDao
import com.qadam.lastpuff.data.local.dao.UserProfileDao
import com.qadam.lastpuff.data.local.entity.RelapseEventEntity
import com.qadam.lastpuff.data.mapper.toDomain
import com.qadam.lastpuff.data.mapper.toEntity
import com.qadam.lastpuff.domain.model.Achievement
import com.qadam.lastpuff.domain.model.CravingEvent
import com.qadam.lastpuff.domain.model.HealthMilestone
import com.qadam.lastpuff.domain.model.HomeStats
import com.qadam.lastpuff.domain.model.MoneyGoal
import com.qadam.lastpuff.domain.model.ProgressStats
import com.qadam.lastpuff.domain.model.RecoveryBoosts
import com.qadam.lastpuff.domain.model.RecoveryIndex
import com.qadam.lastpuff.domain.model.SosContact
import com.qadam.lastpuff.domain.model.UserProfile
import com.qadam.lastpuff.domain.model.VictoryRecord
import com.qadam.lastpuff.domain.support.MessageSession
import com.qadam.lastpuff.domain.support.SupportMessageBank
import com.qadam.lastpuff.util.AppConstants
import com.qadam.lastpuff.util.RecoveryCalculator
import com.qadam.lastpuff.util.StatsCalculator
import com.qadam.lastpuff.util.TimeFormatUtils
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map

class UserRepository(
    private val userProfileDao: UserProfileDao,
    private val cravingEventDao: CravingEventDao,
    private val relapseEventDao: RelapseEventDao,
    private val moneyGoalDao: MoneyGoalDao,
    private val sosContactDao: SosContactDao,
    private val preferencesManager: PreferencesManager
) {
    val onboardingCompleted: Flow<Boolean> = preferencesManager.onboardingCompleted
    val darkTheme: Flow<Boolean?> = preferencesManager.darkTheme
    val notificationsEnabled: Flow<Boolean> = preferencesManager.notificationsEnabled
    val notificationHour: Flow<Int> = preferencesManager.notificationHour
    val notificationMinute: Flow<Int> = preferencesManager.notificationMinute
    val familyPhotoUri: Flow<String?> = preferencesManager.familyPhotoUri
    val personalLetter: Flow<String?> = preferencesManager.personalLetter

    fun observeProfile(): Flow<UserProfile?> =
        userProfileDao.observeProfile().map { it?.toDomain() }

    fun observeSosContact(): Flow<SosContact?> =
        sosContactDao.observeContact().map { it?.toDomain() }

    fun observeMoneyGoal(): Flow<MoneyGoal?> =
        moneyGoalDao.observeActiveGoal().map { it?.toDomain() }

    fun observeCravingEvents(): Flow<List<CravingEvent>> =
        cravingEventDao.observeAll().map { list -> list.map { it.toDomain() } }

    suspend fun getProfile(): UserProfile? = userProfileDao.getProfile()?.toDomain()

    suspend fun saveProfile(profile: UserProfile) {
        userProfileDao.insert(profile.toEntity())
    }

    suspend fun completeOnboarding(
        profile: UserProfile,
        sosContact: SosContact,
        personalLetter: String?
    ) {
        userProfileDao.insert(profile.toEntity())
        sosContactDao.insert(sosContact.toEntity())
        preferencesManager.setPersonalLetter(personalLetter)
        preferencesManager.setOnboardingCompleted(true)
    }

    suspend fun saveSosContact(contact: SosContact) {
        val existing = sosContactDao.observeContact().first()
        if (existing != null) {
            sosContactDao.update(contact.copy(id = existing.id).toEntity())
        } else {
            sosContactDao.insert(contact.toEntity())
        }
    }

    suspend fun saveMoneyGoal(goal: MoneyGoal) {
        val existing = moneyGoalDao.observeActiveGoal().first()
        if (existing != null) {
            moneyGoalDao.update(goal.copy(id = existing.id).toEntity())
        } else {
            moneyGoalDao.insert(goal.toEntity())
        }
    }

    suspend fun recordCraving(intensity: Int, trigger: String, success: Boolean) {
        val event = CravingEvent(
            createdAt = System.currentTimeMillis(),
            intensity = intensity,
            trigger = trigger,
            success = success
        )
        cravingEventDao.insert(event.toEntity())

        if (!success) {
            relapseEventDao.insert(
                RelapseEventEntity(createdAt = System.currentTimeMillis())
            )
            val profile = getProfile() ?: return
            userProfileDao.update(
                profile.copy(lastSmokeDate = System.currentTimeMillis()).toEntity()
            )
        } else {
            preferencesManager.addCoin()
        }

        checkAndUnlockAchievements()
    }

    suspend fun applySosRecoveryBoost(): Pair<RecoveryIndex, RecoveryIndex> {
        val profile = getProfile() ?: return RecoveryIndex() to RecoveryIndex()
        val cravings = cravingEventDao.getAll().map { it.toDomain() }
        val now = System.currentTimeMillis()
        val elapsed = StatsCalculator.elapsedMillis(profile.lastSmokeDate, now)
        val wins = cravings.count { it.success }
        val before = computeRecoveryIndex(elapsed, wins)
        val after = (before + RecoveryBoosts.SOS_WIN).clamp()
        preferencesManager.setRecoveryIndex(RecoveryCalculator.encode(after))
        return before to after
    }

    fun observeRecoveryIndex(
        nowFlow: Flow<Long> = kotlinx.coroutines.flow.flowOf(System.currentTimeMillis())
    ): Flow<RecoveryIndex> = combine(
        preferencesManager.recoveryIndexRaw,
        observeProfile(),
        observeCravingEvents(),
        nowFlow
    ) { raw, profile, cravings, now ->
        val profileData = profile ?: return@combine RecoveryIndex()
        val elapsed = StatsCalculator.elapsedMillis(profileData.lastSmokeDate, now)
        val wins = cravings.count { it.success }
        computeRecoveryIndexFromRaw(raw, elapsed, wins)
    }

    private fun computeRecoveryIndexFromRaw(raw: String?, elapsed: Long, wins: Int): RecoveryIndex {
        val stored = RecoveryCalculator.decode(raw)
        val timeBased = RecoveryCalculator.fromTimeWithoutSmoking(elapsed, wins)
        return RecoveryCalculator.merge(stored, timeBased)
    }

    private suspend fun computeRecoveryIndex(elapsed: Long, wins: Int): RecoveryIndex {
        val raw = preferencesManager.recoveryIndexRaw.first()
        return computeRecoveryIndexFromRaw(raw, elapsed, wins)
    }

    suspend fun updateLastSmokeDate(timestamp: Long) {
        val profile = getProfile() ?: return
        userProfileDao.update(profile.copy(lastSmokeDate = timestamp).toEntity())
        checkAndUnlockAchievements()
    }

    suspend fun resetProgress() {
        cravingEventDao.deleteAll()
        relapseEventDao.deleteAll()
        val profile = getProfile()
        if (profile != null) {
            userProfileDao.update(
                profile.copy(lastSmokeDate = System.currentTimeMillis()).toEntity()
            )
        }
        preferencesManager.clearAll()
        preferencesManager.setOnboardingCompleted(true)
    }

    suspend fun setDarkTheme(enabled: Boolean) = preferencesManager.setDarkTheme(enabled)
    suspend fun setNotificationsEnabled(enabled: Boolean) = preferencesManager.setNotificationsEnabled(enabled)
    suspend fun setNotificationTime(hour: Int, minute: Int) = preferencesManager.setNotificationTime(hour, minute)
    suspend fun setPersonalLetter(letter: String?) = preferencesManager.setPersonalLetter(letter)
    suspend fun setFamilyPhotoUri(uri: String?) = preferencesManager.setFamilyPhotoUri(uri)

    fun observeHomeStats(nowFlow: Flow<Long> = kotlinx.coroutines.flow.flowOf(System.currentTimeMillis())): Flow<HomeStats?> =
        combine(
            combine(
                observeProfile(),
                observeCravingEvents(),
                observeMoneyGoal()
            ) { profile, cravings, goal -> Triple(profile, cravings, goal) },
            combine(
                preferencesManager.unlockedAchievementsFlow(),
                preferencesManager.totalCoinsFlow(),
                nowFlow
            ) { unlocked, coins, now -> Triple(unlocked, coins, now) }
        ) { data, meta ->
            val (profile, cravings, goal) = data
            val (unlocked, coins, now) = meta
            profile ?: return@combine null
            val dayIndex = (now / (24 * 60 * 60 * 1000L)).toInt()
            val hour = java.util.Calendar.getInstance().apply { timeInMillis = now }
                .get(java.util.Calendar.HOUR_OF_DAY)
            val daySession = MessageSession(dayIndex)
            val quote = SupportMessageBank.dayMessage(hour, daySession)
            val achievements = buildAchievements(profile, unlocked, cravings)
            val nextAchievement = achievements.firstOrNull { !it.isUnlocked }
            val days = StatsCalculator.daysWithoutSmoking(profile.lastSmokeDate, now)
            val hours = StatsCalculator.hoursWithoutSmoking(profile.lastSmokeDate, now) % 24
            val lastWin = cravings.firstOrNull { it.success }

            HomeStats(
                days = days,
                hours = hours,
                minutes = StatsCalculator.minutesWithoutSmoking(profile.lastSmokeDate, now),
                cigarettesNotSmoked = StatsCalculator.cigarettesNotSmoked(profile, now),
                moneySaved = StatsCalculator.moneySaved(profile, now),
                motivationalQuote = quote,
                nextAchievement = nextAchievement,
                lastVictoryAgo = lastWin?.let { TimeFormatUtils.formatAgo(it.createdAt, now) },
                dailyLifeCard = AppConstants.dailyLifeCard(days, hours),
                personalReason = SupportMessageBank.memoryForReason(
                    profile.reasons.firstOrNull() ?: "",
                    goal?.title
                ) ?: AppConstants.personalReasonMessage(profile.reasons, goal?.title),
                totalCoins = coins,
                isJustStarted = days == 0 && hours < 1
            )
        }

    fun observeVictories(): Flow<List<VictoryRecord>> =
        observeCravingEvents().map { events ->
            events.filter { it.success }.map { event ->
                VictoryRecord(
                    id = event.id,
                    time = TimeFormatUtils.formatTime(event.createdAt),
                    trigger = event.trigger.ifBlank { "Тяга" },
                    intensity = event.intensity,
                    message = when {
                        event.intensity >= 8 -> "Очень хотелось. Справился."
                        event.trigger.isNotBlank() -> "После «${event.trigger.lowercase()}». Не закурил."
                        else -> "Справился с тягой."
                    }
                )
            }
        }

    fun observeProgressStats(nowFlow: Flow<Long> = kotlinx.coroutines.flow.flowOf(System.currentTimeMillis())): Flow<ProgressStats?> =
        combine(
            observeProfile(),
            observeCravingEvents(),
            relapseEventDao.observeAll(),
            preferencesManager.longestStreakFlow(),
            nowFlow
        ) { profile, cravings, relapses, storedStreak, now ->
            profile ?: return@combine null
            val days = StatsCalculator.daysWithoutSmoking(profile.lastSmokeDate, now)
            val longestStreak = maxOf(days, storedStreak)
            val wins = cravings.count { it.success }
            val relapseCount = relapses.size

            ProgressStats(
                daysWithoutSmoking = days,
                hoursWithoutSmoking = StatsCalculator.hoursWithoutSmoking(profile.lastSmokeDate, now),
                cigarettesNotSmoked = StatsCalculator.cigarettesNotSmoked(profile, now),
                moneySaved = StatsCalculator.moneySaved(profile, now),
                totalCravings = cravings.size,
                cravingsWon = wins,
                relapses = relapseCount,
                winRate = StatsCalculator.winRate(wins, cravings.size),
                longestStreak = longestStreak,
                dangerousHours = StatsCalculator.dangerousHours(cravings.map { it.createdAt }),
                topTriggers = StatsCalculator.topTriggers(cravings.map { it.trigger }),
                averageIntensity = StatsCalculator.averageIntensity(cravings.map { it.intensity })
            )
        }

    fun observeHealthMilestones(nowFlow: Flow<Long> = kotlinx.coroutines.flow.flowOf(System.currentTimeMillis())): Flow<List<HealthMilestone>> =
        combine(observeProfile(), nowFlow) { profile, now ->
            val lastSmoke = profile?.lastSmokeDate ?: return@combine emptyList()
            val elapsed = StatsCalculator.elapsedMillis(lastSmoke, now)
            AppConstants.HEALTH_MILESTONES.map { (title, desc, duration) ->
                HealthMilestone(
                    id = title,
                    title = title,
                    description = desc,
                    durationMillis = duration,
                    isUnlocked = elapsed >= duration
                )
            }
        }

    fun observeAchievements(): Flow<List<Achievement>> =
        combine(observeProfile(), observeCravingEvents(), preferencesManager.unlockedAchievementsFlow()) { profile, cravings, unlocked ->
            if (profile == null) emptyList()
            else buildAchievements(profile, unlocked, cravings)
        }

    private suspend fun checkAndUnlockAchievements() {
        val profile = getProfile() ?: return
        val cravings = cravingEventDao.getAll().map { it.toDomain() }
        val unlocked = preferencesManager.unlockedAchievementsFlow().first()
        val achievements = buildAchievements(profile, unlocked, cravings)
        achievements.filter { it.isUnlocked && it.id !in unlocked }.forEach {
            preferencesManager.unlockAchievement(it.id)
        }
    }

    private fun buildAchievements(
        profile: UserProfile,
        unlocked: Set<String>,
        cravings: List<CravingEvent> = emptyList()
    ): List<Achievement> {
        val now = System.currentTimeMillis()
        val elapsed = StatsCalculator.elapsedMillis(profile.lastSmokeDate, now)
        val days = StatsCalculator.daysWithoutSmoking(profile.lastSmokeDate, now)
        val cigarettes = StatsCalculator.cigarettesNotSmoked(profile, now)
        val money = StatsCalculator.moneySaved(profile, now)
        val wins = cravings.count { it.success }
        val hasRelapse = cravings.any { !it.success }

        val conditions = mapOf(
            "first_hour" to (elapsed >= 60 * 60 * 1000L),
            "first_day" to (days >= 1),
            "three_days" to (days >= 3),
            "seven_days" to (days >= 7),
            "thirty_days" to (days >= 30),
            "hundred_days" to (days >= 100),
            "cigarettes_100" to (cigarettes >= 100),
            "money_10000" to (money >= 10000),
            "wins_10" to (wins >= 10),
            "wins_50" to (wins >= 50),
            "relapse_survived" to hasRelapse
        )

        return AppConstants.ACHIEVEMENTS.map { (id, title, desc) ->
            val isUnlocked = id in unlocked || conditions[id] == true
            Achievement(id = id, title = title, description = desc, isUnlocked = isUnlocked)
        }
    }
}

File: ios/QadamLastPuff/Data/UserRepository.swift
import Foundation
import Combine

@MainActor
final class UserRepository: ObservableObject {
    let store: DataStore
    let preferences: PreferencesManager
    private var cancellables = Set<AnyCancellable>()

    init(store: DataStore, preferences: PreferencesManager) {
        self.store = store
        self.preferences = preferences
        syncLaunchState()
        store.objectWillChange
            .receive(on: RunLoop.main)
            .sink { [weak self] _ in self?.objectWillChange.send() }
            .store(in: &cancellables)
    }

    var onboardingCompleted: Bool {
        get { preferences.onboardingCompleted }
        set { preferences.onboardingCompleted = newValue }
    }

    var isFirstLaunch: Bool {
        get { preferences.isFirstLaunch }
        set { preferences.isFirstLaunch = newValue }
    }

    var profile: UserProfile? { store.data.profile }
    var sosContact: SosContact? { store.data.sosContact }
    var moneyGoal: MoneyGoal? { store.data.moneyGoal }
    var completedGoals: [CompletedMoneyGoal] { store.data.completedGoals }
    var cravings: [CravingEvent] { store.data.cravings }
    var relapses: [RelapseEvent] { store.data.relapses }

    /// Синхронизирует флаги запуска с реальными данными профиля.
    func syncLaunchState() {
        if profile != nil {
            if !preferences.onboardingCompleted {
                preferences.onboardingCompleted = true
            }
        } else if preferences.onboardingCompleted {
            // Флаг есть, а профиля нет — сбрасываем, чтобы снова показать onboarding.
            preferences.onboardingCompleted = false
        }
    }

    func saveProfile(_ profile: UserProfile) {
        store.update { $0.profile = profile }
        checkAndUnlockAchievements()
    }

    func completeOnboarding(profile: UserProfile, sosContact: SosContact, personalLetter: String?) {
        store.update {
            $0.profile = profile
            $0.sosContact = sosContact
        }
        preferences.personalLetter = personalLetter
        preferences.onboardingCompleted = true
    }

    func saveSosContact(_ contact: SosContact) {
        store.update { data in
            var c = contact
            if let existing = data.sosContact { c.id = existing.id }
            data.sosContact = c
        }
    }

    func saveMoneyGoal(_ goal: MoneyGoal) {
        store.update { data in
            var g = goal
            if let existing = data.moneyGoal { g.id = existing.id }
            data.moneyGoal = g
        }
        checkGoalReached()
    }

    func archiveCompletedGoal() {
        guard let profile, let goal = moneyGoal else { return }
        let saved = StatsCalculator.moneySaved(profile: profile)
        guard saved >= goal.amount else { return }
        store.update { data in
            let completed = CompletedMoneyGoal(
                id: data.nextCompletedGoalId,
                title: goal.title,
                amount: goal.amount,
                savedAmount: saved,
                completedAt: Date()
            )
            data.nextCompletedGoalId += 1
            data.completedGoals.insert(completed, at: 0)
            data.moneyGoal = nil
        }
    }

    func checkGoalReached() -> Bool {
        guard let profile, let goal = moneyGoal else { return false }
        return StatsCalculator.moneySaved(profile: profile) >= goal.amount
    }

    func moneySavedNow() -> Double {
        guard let profile else { return 0 }
        return StatsCalculator.moneySaved(profile: profile)
    }

    @discardableResult
    func recordCraving(intensity: Int, trigger: String, location: String, success: Bool) -> Int {
        store.update { data in
            let event = CravingEvent(
                id: data.nextCravingId,
                createdAt: Date(),
                intensity: intensity,
                trigger: trigger,
                location: location,
                success: success
            )
            data.nextCravingId += 1
            data.cravings.insert(event, at: 0)

            if !success {
                let relapse = RelapseEvent(id: data.nextRelapseId, createdAt: Date())
                data.nextRelapseId += 1
                data.relapses.insert(relapse, at: 0)
                if var profile = data.profile {
                    profile.lastSmokeDate = Date()
                    data.profile = profile
                }
            }
        }
        if success {
            let coins = AppConstants.coinReward(for: intensity)
            preferences.addCoins(coins)
            checkAndUnlockAchievements()
            return coins
        }
        checkAndUnlockAchievements()
        return 0
    }

    func recordRelapse() {
        recordCraving(intensity: 5, trigger: "Срыв", location: "", success: false)
    }

    func applySosRecoveryBoost() -> (RecoveryIndex, RecoveryIndex) {
        guard let profile else { return (RecoveryIndex(), RecoveryIndex()) }
        let now = Date()
        let elapsed = StatsCalculator.elapsedMillis(lastSmokeDate: profile.lastSmokeDate, now: now)
        let wins = cravings.filter(\.success).count
        let before = computeRecoveryIndex(elapsed: elapsed, wins: wins)
        let after = (before + RecoveryBoosts.sosWin).clamped()
        preferences.recoveryIndexRaw = RecoveryCalculator.encode(after)
        return (before, after)
    }

    func updateLastSmokeDate(_ date: Date) {
        guard var p = profile else { return }
        p.lastSmokeDate = date
        saveProfile(p)
    }

    func resetProgress() {
        store.update { data in
            data.cravings = []
            data.relapses = []
            if var profile = data.profile {
                profile.lastSmokeDate = Date()
                data.profile = profile
            }
        }
        preferences.clearAll()
        preferences.onboardingCompleted = true
    }

    func computeRecoveryIndex(now: Date = Date()) -> RecoveryIndex {
        guard let profile else { return RecoveryIndex() }
        let elapsed = StatsCalculator.elapsedMillis(lastSmokeDate: profile.lastSmokeDate, now: now)
        let wins = cravings.filter(\.success).count
        return computeRecoveryIndex(elapsed: elapsed, wins: wins)
    }

    private func computeRecoveryIndex(elapsed: Int64, wins: Int) -> RecoveryIndex {
        let stored = RecoveryCalculator.decode(preferences.recoveryIndexRaw)
        let timeBased = RecoveryCalculator.fromTimeWithoutSmoking(elapsedMs: elapsed, wins: wins)
        return RecoveryCalculator.merge(stored: stored, timeBased: timeBased)
    }

    func buildHomeStats(now: Date = Date()) -> HomeStats? {
        guard let profile else { return nil }
        let dayIndex = Int(now.timeIntervalSince1970 / 86400)
        let hour = Calendar.current.component(.hour, from: now)
        let session = MessageSession(seed: dayIndex)
        let quote = SupportMessageBank.dayMessage(hour: hour, session: session)
        let achievements = buildAchievements(now: now)
        let nextAchievement = achievements.first { !$0.isUnlocked }
        let days = StatsCalculator.daysWithoutSmoking(lastSmokeDate: profile.lastSmokeDate, now: now)
        let hours = StatsCalculator.hoursWithoutSmoking(lastSmokeDate: profile.lastSmokeDate, now: now) % 24
        let lastWin = cravings.first { $0.success }

        return HomeStats(
            days: days,
            hours: hours,
            minutes: StatsCalculator.minutesWithoutSmoking(lastSmokeDate: profile.lastSmokeDate, now: now),
            cigarettesNotSmoked: StatsCalculator.cigarettesNotSmoked(profile: profile, now: now),
            moneySaved: StatsCalculator.moneySaved(profile: profile, now: now),
            motivationalQuote: quote,
            nextAchievement: nextAchievement,
            lastVictoryAgo: lastWin.map { TimeFormatUtils.formatAgo($0.createdAt, now: now) },
            dailyLifeCard: AppConstants.dailyLifeCard(days: days, hours: hours),
            personalReason: SupportMessageBank.memoryForReason(profile.reasons.first ?? "", goalTitle: moneyGoal?.title)
                ?? AppConstants.personalReasonMessage(reasons: profile.reasons, goalTitle: moneyGoal?.title),
            totalCoins: preferences.totalCoins,
            isJustStarted: days == 0 && hours < 1
        )
    }

    func buildProgressStats(now: Date = Date()) -> ProgressStats? {
        guard let profile else { return nil }
        let days = StatsCalculator.daysWithoutSmoking(lastSmokeDate: profile.lastSmokeDate, now: now)
        let longestStreak = max(days, preferences.longestStreak)
        if days > preferences.longestStreak { preferences.longestStreak = days }
        let wins = cravings.filter(\.success).count

        return ProgressStats(
            daysWithoutSmoking: days,
            hoursWithoutSmoking: StatsCalculator.hoursWithoutSmoking(lastSmokeDate: profile.lastSmokeDate, now: now),
            cigarettesNotSmoked: StatsCalculator.cigarettesNotSmoked(profile: profile, now: now),
            moneySaved: StatsCalculator.moneySaved(profile: profile, now: now),
            totalCravings: cravings.count,
            cravingsWon: wins,
            relapses: relapses.count,
            winRate: StatsCalculator.winRate(wins: wins, total: cravings.count),
            longestStreak: longestStreak,
            dangerousHours: StatsCalculator.dangerousHours(events: cravings.map(\.createdAt)),
            topTriggers: StatsCalculator.topTriggers(triggers: cravings.map(\.trigger)),
            averageIntensity: StatsCalculator.averageIntensity(intensities: cravings.map(\.intensity))
        )
    }

    func buildVictories() -> [VictoryRecord] {
        cravings.filter(\.success).map { event in
            let message: String
            if event.intensity >= 8 {
                message = "Очень хотелось. Справился."
            } else if !event.trigger.isEmpty {
                message = "После «\(event.trigger.lowercased())». Не закурил."
            } else {
                message = "Справился с тягой."
            }
            return VictoryRecord(
                id: event.id,
                time: TimeFormatUtils.formatTime(event.createdAt),
                trigger: event.trigger.isEmpty ? "Тяга" : event.trigger,
                intensity: event.intensity,
                message: message
            )
        }
    }

    func buildHealthMilestones(now: Date = Date()) -> [HealthMilestone] {
        guard let profile else { return [] }
        let elapsed = StatsCalculator.elapsedMillis(lastSmokeDate: profile.lastSmokeDate, now: now)
        return AppConstants.healthMilestones.map { title, desc, duration in
            HealthMilestone(id: title, title: title, description: desc, durationMillis: duration, isUnlocked: elapsed >= duration)
        }
    }

    func buildAchievements(now: Date = Date()) -> [Achievement] {
        guard let profile else {
            return AppConstants.achievements.map { id, title, desc in
                Achievement(id: id, title: title, description: desc, isUnlocked: false)
            }
        }
        let unlocked = preferences.unlockedAchievements
        let elapsed = StatsCalculator.elapsedMillis(lastSmokeDate: profile.lastSmokeDate, now: now)
        let days = StatsCalculator.daysWithoutSmoking(lastSmokeDate: profile.lastSmokeDate, now: now)
        let cigarettes = StatsCalculator.cigarettesNotSmoked(profile: profile, now: now)
        let money = StatsCalculator.moneySaved(profile: profile, now: now)
        let wins = cravings.filter(\.success).count
        let hasRelapse = cravings.contains { !$0.success }
        let totalCoins = preferences.totalCoins

        let conditions: [String: Bool] = [
            "first_hour": elapsed >= 60 * 60 * 1000,
            "first_day": days >= 1,
            "three_days": days >= 3,
            "seven_days": days >= 7,
            "thirty_days": days >= 30,
            "hundred_days": days >= 100,
            "cigarettes_100": cigarettes >= 100,
            "money_10000": money >= 10000,
            "wins_10": wins >= 10,
            "wins_50": wins >= 50,
            "coins_20": totalCoins >= 20,
            "coins_40": totalCoins >= 40,
            "coins_80": totalCoins >= 80,
            "coins_110": totalCoins >= 110,
            "relapse_survived": hasRelapse
        ]

        return AppConstants.achievements.map { id, title, desc in
            Achievement(id: id, title: title, description: desc, isUnlocked: unlocked.contains(id) || (conditions[id] == true))
        }
    }

    private func checkAndUnlockAchievements() {
        guard let profile else { return }
        let achievements = buildAchievements()
        let unlocked = preferences.unlockedAchievements
        for achievement in achievements where achievement.isUnlocked && !unlocked.contains(achievement.id) {
            preferences.unlockAchievement(achievement.id)
        }
        _ = profile
    }
}
